src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt

```diff
@@ -12,237 +12,275 @@ import kotlinx.coroutines.flow.*
-                logger.info { "connection state changed: ${it::class.simpleName}" }
-                _connectionState.value = it
-                previousState = it
-            }
+    data class ListenJob(val job: Job, val socket: TcpSocket) {
+        fun cancel() {
+            job.cancel()
+            socket.close()
         }
     }
 
-    fun connect(serverAddress: ServerAddress) {
+    private var listenJob: ListenJob? = null
+
+    suspend fun connect(serverAddress: ServerAddress, socketBuilder: TcpSocket.Builder, timeout: Duration = 15.seconds): Boolean {
         if (_connectionStatus.value is ElectrumConnectionStatus.Closed) {
-            runJob = establishConnection(serverAddress)
-        } else logger.warning { "electrum client is already running" }
+            listenJob?.cancel()
+            val socket = openSocket(serverAddress, socketBuilder, timeout) ?: return false
+            logger.info { "connected to electrumx instance" }
+            return try {
+                handshake(socket, timeout)
+                listenJob = listen(socket)
+                true
+            } catch (ex: Throwable) {
+                logger.warning(ex) { "electrum connection handshake failed: " }
+                val ioException = when (ex) {
+                    is TcpSocket.IOException -> ex
+                    else -> TcpSocket.IOException.Unknown(ex.message, ex)
+                }
+                socket.close()
+                _connectionStatus.value = ElectrumConnectionStatus.Closed(ioException)
+                false
+            }
+        } else {
+            logger.warning { "ignoring connection request, electrum client is already running" }
+            return false
+        }
     }
 
-    @OptIn(DelicateCoroutinesApi::class)
     fun disconnect() {
-        launch {
-            if (!mailbox.isClosedForSend) {
-                mailbox.send(Action.Disconnect)
-            }
-        }
+        listenJob?.cancel()
+        listenJob = null
+        _connectionStatus.value = ElectrumConnectionStatus.Closed(null)
     }
 
-    private fun establishConnection(serverAddress: ServerAddress) = launch(CoroutineExceptionHandler { _, exception ->
-        logger.error(exception) { "error starting electrum client: " }
-    }) {
-        _connectionStatus.value = ElectrumConnectionStatus.Connecting
-        val socket: TcpSocket = try {
+    private suspend fun openSocket(serverAddress: ServerAddress, socketBuilder: TcpSocket.Builder, timeout: Duration): TcpSocket? {
+        return try {
+            _connectionStatus.value = ElectrumConnectionStatus.Connecting
             val (host, port, tls) = serverAddress
             logger.info { "attempting connection to electrumx instance [host=$host, port=$port, tls=$tls]" }
-            socketBuilder?.connect(host, port, tls, loggerFactory) ?: error("socket builder is null.")
+            withTimeout(timeout) {
+                socketBuilder.connect(host, port, tls, loggerFactory)
```

Comment:
Interesting, good catch! I don't see how we can fix this though? I started writing the following change:

```kotlin
var socket: TcpSocket? = null
try {
    ...
    withTimeout(timeout) {
        socket = socketBuilder.connect(host, port, tls, loggerFactory)
        socket
    }
} catch (ex: Throwable) {
    socket?.close()
}
```

But it doesn't actually get rid of the resource leak, because the timeout exception could be thrown after establishing the tcp socket but before we assign it to our `var`. I don't see a good way to fix this, do you have an idea or should we just ignore it because it's very unlikely and not a large resource leak anyway?

Reply:
this seems to be the fix suggested by Jetbrains, so I guess we should go for it (even though such a leak is very unlikely)

Reply:
Done in https://github.com/ACINQ/lightning-kmp/pull/512/commits/4502d8619c63daa25bed8a95f1d90310f2dc74ce